feat(subgraph events): emit SubgraphEvent for subgraph changes

aaronArinder opened the pull request:
- with an enum, we can capture different events more fine-grained
  - includes SubgraphRemoved as an event

aaronArinder committed:
feat(subgraph events): emit SubgraphEvent for subgraph changes
  - with an enum, we can capture different events more fine-grained
  - includes SubgraphRemoved as an event

dotdat commented on src/composition/runner.rs:
```diff
@@ -194,6 +194,15 @@ impl SubgraphWatchers {
+/// name of the subgraph
+pub enum SubgraphEvent {
+    /// A change to the watched subgraph
+    SubgraphChanged(SubgraphChanged),
```
I think this should probably be `SubgraphSchemaChanged`. `SubgraphWatchers` should be watching the `SupergraphConfig` file for changes and informing composition that the schema was either changed or removed.

Subgraphs being added amounts to a SubgraphSchemaChanged event for the purpose of composition, as it is just upserting sdls into its internal BTreeMap.

aaronArinder replied:
I think we're on the same page; the watchers emit `SubgraphSchemaChanged` events, and in the tokio tasks we spin up for those watches, we send a `SubgraphChanged` event to make sure we're capturing the different kinds of events that can trigger recomposition (including subgraph removals--there's no SDL to emit in that case, and instead of marking it optional we're emitting a different event, `SubgraphRemoved`)

aaronArinder replied:
(sorted elsewhere; should reflect the design now)

aaronArinder committed:
feat(subgraph watching): support SDL, SchemaSource::Subgraph changes (#2209)

```    
feat(subgraph watching): support SDL, SchemaSource::Subgraph changes
      - when the supergraphconfig has changes to its in-lined SDL or
        subgraph subgraphs (I hate that it's called this because it's so
        confusing), we now have a way of seeing what's changed (rather than
        just added or removed)
      - we no longer use fed-rs's get_subgraph_defintions() because it'll
        error without fully resolved configs (ie, evertying already
        in-lined; for SchemaSource::Subgraph, we won't have that)
      - refetches SchemaSource::Subgraphs when their graphref or subgraph
        name change
      - when a SchemaSource::SDL has its sdl change, we emit that new sdl as
        a change
```

dotdat commented on src/composition/runner.rs:
```diff
@@ -243,40 +267,101 @@ impl SubtaskHandleStream for SubgraphWatchers {
+                                Subtask::<SubgraphWatcher, WatchedSdlChange>::new(subgraph_watcher);
+
+                            let sender = sender.clone();
+                            let subgraph_name_c = subgraph_name.clone();
```
nit: `subtask_clone_c` feels like a bit of a smell. It might feel slightly cleaner if we reused the subgraph_name variable in a block (which seems to be an idiomatic rust thing). So like:
```
let messages_abort_handle = {
  let sender = sender.clone();
  let subgraph_name = subgraph_name.clone();
  tokio::spawn({
    ...
  }).abort_handle();
}
```

dotdat commented on src/composition/run_composition.rs:
```diff
@@ -143,8 +143,9 @@ mod tests {
 
-        let subgraph_change_events: BoxStream<SubgraphChanged> =
-            once(async { SubgraphChanged::default() }).boxed();
+        let subgraph_change_events: BoxStream<SubgraphEvent> =
```
I'm not sure we even need this/want this? I'm not sure what the value is of producing an empty event.
